sys/pool_alloc.c

```diff
@@ -0,0 +1,235 @@
+  pool->pp_itemsize = size;
+  pool->pp_nslabs = 1;
+  pool->pp_nitems = first_slab->ph_ntotal;
+  PALLOC_DEBUG_PRINT("Initialized new pool at 0x%lx\n", (unsigned long)pool);
+}
+
+void pool_destroy(pool_t *pool) {
+  /*
+  destroy the objects;
+  free the underlying memory;
+  */
+
+  PALLOC_DEBUG_PRINT("Entering pool_destroy\n");
+
+  unsigned long *tmp = (unsigned long *)pool;
+
+  for (size_t i = 0; i < sizeof(pool_t) / sizeof(unsigned long); i++) {
+    if (tmp[i] == PP_FREED_WORD) {
+      panic("double free at pool 0x%lx\n", (unsigned long)pool);
+    }
+  }
+
+  PALLOC_DEBUG_PRINT("Destroying empty slabs\n");
+  pool_slab_t *it = LIST_FIRST(&pool->pp_empty_slabs);
+  while (it) {
+    pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+    LIST_REMOVE(it, ph_slablist);
+    destroy_slab(it, pool);
+    it = next;
+  }
+
+  PALLOC_DEBUG_PRINT("Destroying partially filled slabs\n");
+  it = LIST_FIRST(&pool->pp_part_slabs);
+  while (it) {
+    pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+    LIST_REMOVE(it, ph_slablist);
+    destroy_slab(it, pool);
+    it = next;
+  }
+
+  PALLOC_DEBUG_PRINT("Destroying full slabs\n");
+  it = LIST_FIRST(&pool->pp_full_slabs);
+  while (it) {
+    pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+    LIST_REMOVE(it, ph_slablist);
+    destroy_slab(it, pool);
+    it = next;
+  }
+
+  for (size_t i = 0; i < sizeof(pool_t) / sizeof(unsigned long); i++) {
+    tmp[i] = PP_FREED_WORD;
+  }
+  PALLOC_DEBUG_PRINT("Destroyed pool at 0x%lx\n", (unsigned long)pool);
+}
+
+void *
+pool_alloc(pool_t *pool,
+           __unused uint32_t flags) { /* TODO: implement different flags */
+  /*
+  if (there’s an object in the cache)
+      take it (no construction required);
+  else {
+      allocate memory;
+      construct the object;
+  }
+  */
+  PALLOC_DEBUG_PRINT("Entering pool_alloc\n");
+
+  unsigned long *tmp = (unsigned long *)pool;
+  for (size_t i = 0; i < sizeof(pool_t) / sizeof(unsigned long); i++) {
+    if (tmp[i] == PP_FREED_WORD) {
+      panic("operation on a free pool 0x%lx\n", (unsigned long)pool);
+    }
+  }
+
+  pool_slab_t *slab_to_use;
+  if (pool->pp_nitems) {
+    slab_to_use = LIST_EMPTY(&pool->pp_part_slabs)
+                    ? LIST_FIRST(&pool->pp_empty_slabs)
+                    : LIST_FIRST(&pool->pp_part_slabs);
+    LIST_REMOVE(slab_to_use, ph_slablist);
+  } else {
+    slab_to_use = create_slab(pool->pp_itemsize, pool, pool->pp_constructor);
+    pool->pp_nitems += slab_to_use->ph_ntotal;
+    pool->pp_nslabs++;
+    PALLOC_DEBUG_PRINT("Growing pool 0x%lx\n", (unsigned long)pool);
+  }
+  void *p = slab_alloc(slab_to_use, pool->pp_itemsize);
+  pool->pp_nitems--;
+  pool_slab_list_t *slab_list_to_insert =
+    slab_to_use->ph_nfree ? &pool->pp_part_slabs : &pool->pp_full_slabs;
+  LIST_INSERT_HEAD(slab_list_to_insert, slab_to_use, ph_slablist);
+
+  return p;
+}
+void pool_free(pool_t *pool,
+               void *ptr) { /* TODO: implement empty slab management */
```

Comment:
Remember that `TODO` comments must stay informative for a long time. Some (decent!) empty slab management is already implemented, so this comment does not tell me what exactly is missing. Imagine it's year 2020, you don't remember any of this code, and someone tasked you with resolving this `TODO` :) Could you make it a bit more verbose?